### ISSUE TITLE
Export ONNX graph with backward pass inside the forward function

### ISSUE THREAD

**Issue by FrancescoMandru** [pytorch/pytorch#88740]
I created a custom model where I'm able to retrieve the activations of a specific layer in order to visually inspect the gradient activations during inference time, this is my example model:

```python
class CustomModel(torch.nn.Module):
    def __init__(self):   
        super().__init__()
        self.model = # Custom model (ResNet, VGG etc etc)
        
        num_ftrs = self.model.classifier[1].in_features
        self.model.classifier[1] = torch.nn.Linear(num_ftrs, 2)
        
        # placeholder for the gradients
        self.gradients = None

    # hook for the gradients of the activations
    def activations_hook(self, grad):
        self.gradients = grad
        
    # method for the gradient extraction
    def get_activations_gradient(self):
        return self.gradients
    
    # method for the activation exctraction
    def get_activations(self, x):
        return self.model.features(x)
    
    # forward function  
    def forward(self, x):
        x = self.model.features(x)
        # register the hook
        if x.requires_grad:
            h = x.register_hook(self.activations_hook)
        
        ### etc etc etc
```

After this, I want to wrap my Model inside a custom one where I compute the [GradCAM](https://arxiv.org/abs/1610.02391) as follows:

```python
class CustomModelWithGradCAM(torch.nn.Module):
    def __init__(self, model_weights, 
                 model_device) -> None:
        super(CustomModelWithGradCAM, self).__init__()  
        
        self.model = CustomModel() # The same as code above already trained
        self.model.load_state_dict(torch.load(model_weights)) # Load its weights
        self.model.to(model_device)
        self.model.eval()
    
        self.model_device = model_device
        
    def getCAM(predictions, classId):
        predictions[:, classId].backward()
        # pull the gradients out of the model
        gradients = self.model.get_activations_gradient()
        ## etc etc etc etc...
     
    def forward(self, x):
        outputs = self.model(x)
        heatmaps = self.getCAM(predictions=outputs, classId=1)
        ## etc etc etc etc...
```

This works very well as I keep the code in Pytorch, however when I try to export this wrapped model in ONNX as follows:

```python
class ONNXCustomModel(torch.nn.Module):
    def __init__(self, weights_path):
        super(ONNXCustomModel, self).__init__()  

        self.model = CustomModelWithGradCAM())
        
    def forward(self, x):
        ..... = self.model(x)
        ## etc etc


#### In the main...

        torch.onnx.export(model,                   # model being run
                        img,                       # model input (or a tuple for multiple inputs)
                        "model.onnx",              # where to save the model (can be a file or file-like object)
                        export_params=True,        # store the trained parameter weights inside the model file
                        opset_version=16,          # the ONNX version to export the model to
                        do_constant_folding=True,  # whether to execute constant folding for optimization
                        operator_export_type=torch.onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK,    #ONNX_ATEN_FALLBACK, the backward ops will be represented as pytorch aten ops
                        input_names = ['input'],   # the model's input names
                        output_names = ["heatmap", "output"], # the model's output names
                        dynamic_axes={'input': {0 : 'batch_size'},    # variable length axes
                                      'output': {0 : 'batch_size'}})

```

It tells me that because of backward pass inside the forward function, ONNX is not able to produce a valid graph, specifically:

> Cannot insert a Tensor that requires grad as a constant. Consider making it a parameter or input, or detaching the gradient

. On the Internet I found that:

> _ONNX export relies on TorchScript, and I believe there's currently no way to capture the backwards graph as TorchScript, so this is not feasible._

I tried a workaround suggested by @thiagocrepaldi who told me:

> if you use operator_export_type=ONNX_ATEN_FALLBACK, the export can succeed and the backward ops will be represented as pytorch aten ops. onnx runtime can undertsand that non-standard-onnx nodes, for instance

but the ONNX export fails. I wonder if there is any trick to do that.

### Versions

```
PyTorch version: 1.12.0+cu116
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.8.15 (default, Oct 12 2022, 19:15:16)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-52-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.5.119
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3070
Nvidia driver version: 510.85.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.2.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.2.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.2.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.2.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.2.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.2.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.2.4
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.3
[pip3] torch==1.12.0+cu116
[pip3] torchmetrics==0.9.3
[pip3] torchvision==0.13.0+cu116
[conda] Could not collect
```

**Comment by AllenTiTaiWang** [comment]
@thiagocrepaldi ATEN_FALL_BACK. Would you take a look?

**Comment by thiagocrepaldi** [comment]
@FrancescoMandru please provide a full repro for this issue.

**Comment by thiagocrepaldi** [comment]
@FrancescoMandru please provide a full repro for this issue? This issue will be closed within 30 days without a repro

**Comment by thiagocrepaldi** [comment]
Closing due to lack of repro script. Please feel free to reopen when a full repro is available
